viscy/data/dynacell.py

```diff
@@ -0,0 +1,224 @@
+                f"prediction database length ({len(self.pred_database)})"
+            )
+
+        # Create datasets
+        datasets = []
+        for i in range(len(self.target_database)):
+            target_data = self.target_database[i]
+            pred_data = self.pred_database[i]
+
+            # Ensure target and prediction metadata match
+            self._validate_matching_metadata(target_data, pred_data, i)
+
+            datasets.append(
+                DynaCellDataset(
+                    cell_type=target_data["cell_type"],
+                    organelle=target_data["organelle"],
+                    infection_condition=target_data["infection_condition"],
+                    pred_dataset=open_ome_zarr(pred_data["zarr_path"]),
+                    target_dataset=open_ome_zarr(target_data["zarr_path"]),
+                    pred_channel=pred_data["channel_name"],
+                    target_channel=target_data["channel_name"],
+                    pred_z_slice=pred_data["z_slice"],
+                    target_z_slice=target_data["z_slice"],
+                )
+            )
+
+        self.test_dataset = ConcatDataset(datasets)
+
+    def _validate_matching_metadata(
+        self, target_data: dict, pred_data: dict, idx: int
+    ) -> None:
+        """Validate that target and prediction metadata match."""
+        # Check cell type
+        if target_data["cell_type"] != pred_data["cell_type"]:
+            raise ValueError(
+                f"Cell type mismatch at index {idx}: "
+                f"target={target_data['cell_type']}, pred={pred_data['cell_type']}"
+            )
+
+        # Check organelle
+        if target_data["organelle"] != pred_data["organelle"]:
+            raise ValueError(
+                f"Organelle mismatch at index {idx}: "
+                f"target={target_data['organelle']}, pred={pred_data['organelle']}"
+            )
+
+        # Check infection condition
+        if target_data["infection_condition"] != pred_data["infection_condition"]:
+            raise ValueError(
+                f"Infection condition mismatch at index {idx}: "
+                f"target={target_data['infection_condition']}, pred={pred_data['infection_condition']}"
+            )
+
+        # Check zarr paths if they should match
+        if target_data["zarr_path"] != pred_data["zarr_path"]:
+            _logger.warning(
+                f"Zarr path mismatch at index {idx}: "
+                f"target={target_data['zarr_path']}, pred={pred_data['zarr_path']}"
+            )
+
+    def test_dataloader(self) -> DataLoader:
+        return DataLoader(
+            self.test_dataset,
+            batch_size=self.batch_size,
+            num_workers=self.num_workers,
+            collate_fn=self._custom_collate,
+        )
+
+    def _custom_collate(self, batch):
+        """Custom collate function that preserves metadata strings."""
+        # Extract metadata from first element in batch
```

Comment:
Is this assuming batch size of 1?

Reply:
yes...

Reply:
Then the constructor should error if it isn't 1.